lib/chef_apply/action/install_chef.rb

```diff
@@ -15,16 +15,111 @@
 # limitations under the License.
 #
 
-require "chef_apply/action/install_chef/base"
-require "chef_apply/action/install_chef/windows"
-require "chef_apply/action/install_chef/linux"
-
-module ChefApply::Action::InstallChef
-  def self.instance_for_target(target_host, opts = { check_only: false })
-    opts[:target_host] = target_host
-    case target_host.base_os
-    when :windows then Windows.new(opts)
-    when :linux then Linux.new(opts)
+require "chef_apply/action/base"
+require "chef_apply/minimum_chef_version"
+require "fileutils"
+
+module ChefApply
+  module Action
+    class InstallChef < ChefApply::Action::Base
+
+      def initialize(opts = { check_only: false })
+        super
+      end
+
+      def perform_action
+        if ChefApply::MinimumChefVersion.check!(target_host, config[:check_only]) == :minimum_version_met
+          notify(:already_installed)
+        else
+          perform_local_install
+        end
+      end
+
+      def name
+        # We have subclasses - so this'll take the qualified name
+        # eg InstallChef::Windows, etc
+        self.class.name.split("::")[-2..-1].join("::")
+      end
+
+      def upgrading?
+        @upgrading
+      end
+
+      def perform_local_install
+        package = lookup_artifact()
```

Comment:
We should add some checking here to append a hash while downloading and then verify the file with it we've had bugs with partial downloads. Not for this pr but capturing it none the less.